### ISSUE TITLE
Incorrect JSON-based Lift

### ISSUE THREAD

**Issue by aaronyoo** [lifting-bits/anvill#8]
I have encountered an issue where `anvill-decompile-json` behaves incorrectly.

C++ code:
```C++
struct Dummy {
    long i;
    long j;
    long k;
};

void dummy_function(struct Dummy& dummy, long i, long j, long k) {
    dummy.i = i; 
    dummy.j = j; 
    dummy.k = k;
}
```

Bitcode from LLVM:
```llvm
; Function Attrs: noinline norecurse nounwind ssp uwtable writeonly
define void @_Z14dummy_functionR5Dummylll(%struct.Dummy* nocapture dereferenceable(24), i64, i64, i64) local_unnamed_addr #5 !dbg !2476 {
  call void @llvm.dbg.value(metadata %struct.Dummy* %0, metadata !2486, metadata !DIExpression()), !dbg !2490
  call void @llvm.dbg.value(metadata i64 %1, metadata !2487, metadata !DIExpression()), !dbg !2491
  call void @llvm.dbg.value(metadata i64 %2, metadata !2488, metadata !DIExpression()), !dbg !2492
  call void @llvm.dbg.value(metadata i64 %3, metadata !2489, metadata !DIExpression()), !dbg !2493
  %5 = getelementptr inbounds %struct.Dummy, %struct.Dummy* %0, i64 0, i32 0, !dbg !2494
  store i64 %1, i64* %5, align 8, !dbg !2495, !tbaa !2496
  %6 = getelementptr inbounds %struct.Dummy, %struct.Dummy* %0, i64 0, i32 1, !dbg !2498
  store i64 %2, i64* %6, align 8, !dbg !2499, !tbaa !2500
  %7 = getelementptr inbounds %struct.Dummy, %struct.Dummy* %0, i64 0, i32 2, !dbg !2501
  store i64 %3, i64* %7, align 8, !dbg !2502, !tbaa !2503
  ret void, !dbg !2504
}
```

JSON Specification:
```json
{
    "arch": "amd64",
    "functions": [
        {
            "name": "dummy_function(Dummy&, long, long, long)",
            "address": 4096,
            "parameters": [
                {
                    "name": "dummy",
                    "register": "RDI",
                    "type": "*{lll}"
                },
                {
                    "name": "i",
                    "register": "RSI",
                    "type": "l"
                },
                {
                    "name": "j",
                    "register": "RDX",
                    "type": "l"
                },
                {
                    "name": "k",
                    "register": "RCX",
                    "type": "l"
                }
            ],
            "return_stack_pointer": {
                "offset": 8,
                "register": "RSP",
                "type": "l"
            },
            "return_address": {
                "memory": {
                    "register": "RSP"
                },
                "type": "L"
            },
            "return_values": []
        }
    ],
    "memory": [
        {
            "address": 4096,
            "data": "554889E54889374889570848894F105DC3",
            "is_readable": true,
            "is_executable": true
        }
    ],
    "stack": {
        "address": 10000,
        "size": 4176,
        "start_offset": 4096
    },
    "os": "macos"
}
```

Lifted Bitcode
```llvm
; Function Attrs: noinline nounwind ssp
define void @"dummy_function(Dummy&, long, long, long)"({ i64, i64, i64 }* nocapture %dummy, i64 %i, i64 %j, i64 %k) local_unnamed_addr #0 {
  %1 = tail call i8* @llvm.returnaddress(i32 0)
  %2 = ptrtoint i8* %1 to i64
  %3 = getelementptr inbounds { i64, i64, i64 }, { i64, i64, i64 }* %dummy, i64 0, i32 0
  store i64 %k, i64* %3, align 8
  %4 = tail call %struct.Memory* @__remill_function_return(%struct.State* nonnull undef, i64 %2, %struct.Memory* null) #3
  ret void
}
```

The lifted bitcode is wrong. One easy way to tell is that there are no offsets and no references to %i or %j. But basically, it is not handling assignment to struct elements correctly.

**Comment by aaronyoo** [comment]
Turns out the reason for this bug is the "convert ptrtoint and adds to geps" optimization which is [here](https://github.com/lifting-bits/anvill/blob/master/lib/Analyze.cpp#L899). In specific, I observed the following behavior (some instructions have been redacted for clarity):

Below is the code before the optimization. We can see that there are 3 memory writes that correspond to the 3 assignments in the original C code:

```llvm
  %5 = ptrtoint { i64, i64, i64 }* %dummy to i64
  %6 = tail call %struct.Memory* @__remill_write_memory_64(%struct.Memory* %4, i64 %5, i64 %i) #4
  %7 = add i64 %5, 8
  %8 = tail call %struct.Memory* @__remill_write_memory_64(%struct.Memory* %6, i64 %7, i64 %j) #4
  %9 = add i64 %5, 16
  %10 = tail call %struct.Memory* @__remill_write_memory_64(%struct.Memory* %8, i64 %9, i64 %k) #4
```

The idea behind `ptrtoint + add  = gep` is that the `ptrtoint + add` pattern is usually used to calculate an address for an array/struct. You convert the pointer to the base of the array/struct to an int and add the correct offset to it. The idea is semantically correct, but there is a bug somewhere. The first symptom of the bug is that the above llvm code is optimized to (I had to add some debug output to get this mid-pass representation):

```
  %12 = ptrtoint { i64, i64, i64 }* %dummy to i64
  %13 = tail call %struct.Memory* @__remill_write_memory_64(%struct.Memory* %11, i64 %12, i64 %i) #4
  %14 = getelementptr { i64, i64, i64 }, { i64, i64, i64 }* %dummy, i64 0
  %15 = ptrtoint { i64, i64, i64 }* %14 to i64
  %16 = add i64 %12, 8
  %17 = tail call %struct.Memory* @__remill_write_memory_64(%struct.Memory* %13, i64 %15, i64 %j) #4
  %18 = getelementptr { i64, i64, i64 }, { i64, i64, i64 }* %dummy, i64 0
  %19 = ptrtoint { i64, i64, i64 }* %18 to i64
  %20 = add i64 %12, 16
  %21 = tail call %struct.Memory* @__remill_write_memory_64(%struct.Memory* %17, i64 %20, i64 %k) #4
```

This is kind of half optimized code so some instructions are dead instructions, but the declaration of `%18` is wrong. It should be accessing the 1 position of the struct, not the 0 position. Herein lies the bug. Basically, the buggy optimization generates bitcode equivalent to:

```C++
dummy.i = i;
dummy.i = j;
dummy.i = k;
```

This gets optimized to: `dummy.i = k` which is why we only see one store instruction in the buggy bitcode. I am working on a fix now.

**Comment by aaronyoo** [comment]
Fixed the above bug by making `indexes` in `GetIndexesInto()` pass by reference. However, this triggered another bug which was fixed by making the type of the index `i32_type`. According to this the [llvm source of CompositeType](https://code.woboq.org/llvm/llvm/lib/IR/Type.cpp.html#551), in order to be a valid index to a structure, the indices need to be of type i32. Previously they were of type i64 but we were getting away with it because we never put more than 1 index into a getelementptr. In the future, all indices to struct type need to be of type i32. Otherwise, `getGEPReturnType()` will try to dereference a nullptr.

Fixed by c3c983d127ae877a253fbb7bc50f153de81f3d68
